staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go

```diff
@@ -141,25 +149,78 @@ func NewApplySet(parent *ApplySetParentRef, tooling ApplySetTooling, mapper meta
+	if a.parentRef.IsNamespaced() {
+		unencoded = strings.Join([]string{a.parentRef.Name, a.parentRef.Namespace, a.parentRef.GroupVersionKind.Kind, a.parentRef.GroupVersionKind.Group}, applySetIDPartDelimiter)
+	} else {
+		unencoded = strings.Join([]string{a.parentRef.Name, a.parentRef.GroupVersionKind.Kind, a.parentRef.GroupVersionKind.Group}, applySetIDPartDelimiter)
+	}
+	w := bytes.Buffer{}
+	encoder := base64.NewEncoder(base64.URLEncoding, &w)
+	_, err := encoder.Write([]byte(unencoded))
+	if err != nil {
+		klog.Fatalf("failed to encode parent ID %s: %w", unencoded, err)
+	}
+	return w.String()
 }
 
 // Validate imposes restrictions on the parent object that is used to track the applyset.
-func (a ApplySet) Validate() error {
+func (a ApplySet) Validate(ctx context.Context, client dynamic.Interface) error {
 	var errors []error
-	// TODO: permit CRDs that have the annotation required by the ApplySet specification
-	if !builtinApplySetParentGVRs[a.parentRef.Resource] {
-		errors = append(errors, fmt.Errorf("resource %q is not permitted as an ApplySet parent", a.parentRef.Resource))
-	}
 	if a.parentRef.IsNamespaced() && a.parentRef.Namespace == "" {
 		errors = append(errors, fmt.Errorf("namespace is required to use namespace-scoped ApplySet"))
 	}
+	if !builtinApplySetParentGVRs.Has(a.parentRef.Resource) {
+		// Determine which custom resource types are allowed as ApplySet parents.
+		// Optimization: Since this makes requests, we only do this if they aren't using a default type.
+		permittedCRParents, err := a.getAllowedCustomResourceParents(ctx, client)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("identifying allowed custom resource parent types: %w", err))
+		}
+		parentRefResourceIgnoreVersion := a.parentRef.Resource.GroupResource().WithVersion("")
+		if !permittedCRParents.Has(parentRefResourceIgnoreVersion) {
+			errors = append(errors, fmt.Errorf("resource %q is not permitted as an ApplySet parent", a.parentRef.Resource))
+		}
+	}
 	return utilerrors.NewAggregate(errors)
 }
 
+func (a *ApplySet) labelForCustomParentCRDs() *metav1.LabelSelector {
+	return &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{{
+			Key:      ApplysetParentCRDLabel,
+			Operator: metav1.LabelSelectorOpExists,
+		}},
+	}
+}
+
+func (a *ApplySet) getAllowedCustomResourceParents(ctx context.Context, client dynamic.Interface) (sets.Set[schema.GroupVersionResource], error) {
+	opts := metav1.ListOptions{
+		LabelSelector: metav1.FormatLabelSelector(a.labelForCustomParentCRDs()),
+	}
+	list, err := client.Resource(schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}).List(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+	set := sets.New[schema.GroupVersionResource]()
+	for i := range list.Items {
+		// Custom resources must be named `<names.plural>.<group>`
+		// and are served under `/apis/<group>/<version>/.../<plural>`
```

Comment:
Is this an acceptable way to get this info? Kubectl does not current import the package for apiextensions types.

Reply:
I _believe_ it is.  Unless someone tells us otherwise, I propose we move forwards with it.